dataflow/src/ops/rewrite.rs

```diff
@@ -0,0 +1,242 @@
+    }
+}
+
+
+impl Ingredient for Rewrite {
+    fn take(&mut self) -> NodeOperator {
+        Clone::clone(self).into()
+    }
+
+    fn ancestors(&self) -> Vec<NodeIndex> {
+        vec![self.src.as_global(), self.signal.as_global()]
+    }
+
+    fn is_join(&self) -> bool {
+        true
+    }
+
+    fn must_replay_among(&self) -> Option<HashSet<NodeIndex>> {
+        Some(Some(self.src.as_global()).into_iter().collect())
+    }
+
+    fn on_connected(&mut self, _: &Graph) {}
+
+    fn on_commit(&mut self, _: NodeIndex, remap: &HashMap<NodeIndex, IndexPair>) {
+        self.src.remap(remap);
+        self.signal.remap(remap);
+    }
+
+    fn on_input(
+        &mut self,
+        from: LocalNodeIndex,
+        rs: Records,
+        _: &mut Tracer,
+        replay_key_col: Option<usize>,
+        nodes: &DomainNodes,
+        state: &StateMap,
+    ) -> ProcessingResult {
+        debug_assert!(from == *self.src || from == *self.signal);
+        let mut misses = Vec::new();
+        let mut emit_rs = Vec::with_capacity(rs.len());
+
+        if rs.is_empty() {
+            return ProcessingResult {
+                results: rs,
+                misses: vec![],
+            };
+        }
+
+        // TODO: it would be nice to just use `rs`,
+        // instead of creating a new `emit_rs`
+        for r in rs {
+            if from == *self.src {
+                let key = r[self.signal_key].clone();
+                // ask signal if column should be rewritten
+                let rc = self.lookup(
+                    *self.signal,
+                    &[0],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+
+                if rc.is_none() {
+                    misses.push(Miss {
+                       node: *self.signal,
+                       columns: vec![0],
+                       replay_key: replay_key_col.map(|col| vec![r[col].clone()]),
+                       key: vec![key],
+                    });
+                }
+
+                let mut rc = rc.unwrap().peekable();
+                if rc.peek().is_none() {
+                    emit_rs.push(r);
+                } else {
+                    let (r, positive) = r.extract();
+                    let row = (self.rewrite(r), positive).into();
+                    emit_rs.push(row);
+                }
+
+            } else if from == *self.signal {
+                let key = r[0].clone();
+                let other_rows = self.lookup(
+                    *self.src,
+                    &[self.signal_key],
+                    &KeyType::Single(&key),
+                    nodes,
+                    state,
+                ).unwrap();
+
+                if other_rows.is_none() {
+                     misses.push(Miss {
+                       node: *self.src,
+                       columns: vec![self.signal_key],
+                       replay_key: replay_key_col.map(|col| vec![r[col].clone()]),
```

Comment:
In fact, I'm fairly sure that `replay_key_col` *must* be `None` in this case, since replays through these nodes always come from the `src` side, right?

Reply:
Yes, they would always come from the source side -- a downstream miss requires data from the `src` side first for sure; however, it may subsequently also require data from the `signal` side to decide whether to rewrite (specifically, if it hasn't seen the key received from `src` before, I think).

Reply:
I think we should check that this is indeed `None` here then, rather than do this (then incorrect) map.